packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,116 @@
+    if (!this.associationMap.get(associationKey)!.has(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptedMessageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? Allowed.NOT_ALLOWED : Allowed.ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.prettyPrintAssociation(scope, portfolio, product)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value, scope.stack.node.addr);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const cfnTagOption = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, cfnTagOption);
+
+          new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+            resourceId: resourceId,
+            tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
+          });
+        }
+      });
+    });
+  }
+
+  private static associationMap = new Map<string, Map<Constraints, CfnConstraintResource>>();
+  private static tagOptionMap = new Map<string, CfnTagOption>();
+
+  private static associationPrecheck(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct): string {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      this.associateProductWithPortfolio(scope, portfolio, product);
+    }
+    return associationKey;
+  }
+
+  private static prettyPrintAssociation(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    return `- Portfolio: ${scope.stack.resolve(portfolio.node.path)} | Product: ${scope.stack.resolve(product.node.path)}`;
+  }
+}
+
+/**
+ * Simplify type for constraint map to hold the L1 constraint resources.
+ */
+type CfnConstraintResource = CfnPortfolioProductAssociation | CfnResourceUpdateConstraint;
+
+/**
+ * Custom allow code
+ */
+enum Allowed {
```

Comment:
Not sure I understand the purpose of this enum...?

Reply:
made it an enum so as not to have strings hardcoded in since it might be a little confusing we use strings instead of booleans, but will remove the enum and just have the strings in the method since they are not used anywhere else.